crates/anvil/src/eth/backend/mem/storage.rs

```diff
@@ -188,6 +192,36 @@ impl InMemoryBlockStates {
             self.disk_cache.remove(on_disk)
         }
     }
+
+    /// Serialize all states to a list of serializable historical states
+    pub fn serialized_states(&mut self) -> SerializableHistoricalStates {
+        // Get in-memory states and serialized them
+        let mut states = self
+            .states
+            .iter_mut()
+            .map(|(hash, state)| (*hash, state.serialize_state()))
+            .collect::<Vec<_>>();
+
+        let disk_states = self
+            .on_disk_states
+            .iter_mut()
+            .map(|(hash, state)| (*hash, state.serialize_state()))
+            .collect::<Vec<_>>();
+
```

Comment:
I think we need to change this function a bit. `serialize_state` does `clear_into_snapshot` internally, thus serializing states basically removes all data from them.

right now it looks like calling `cast rpc "anvil_dumpState" "true"` would result in all historical states being cleared

also, `on_disk_states` contains empty states which are only usable if initialized with a serialized snapshot kept on disk (see `get` fn), so rn we'll just serialize empty states for those that were persisted on disk

Reply:
Makes sense!
I'll add tests for this as well

Reply:
Added a new method to the `MaybeFullDatabase` trait called `read_as_snapshot`. This essentially clones the state into a `StateSnapshot` instead of clearing and moving like `clear_into_snapshot`